### PR TITLE
Capella beacon state

### PR DESCRIPTION
**Pull request by rkapka**
**What type of PR is this?**

Feature

**What does this PR do? Why is it needed?**

Adds in Capella state, including new:
- versions
- types
- getters
- setters
- hashing
- spec params
- SSZ generated code
- proto conversions

This PR introduces the `ExecutionData` interface, which is a common abstraction over `executionPayload`, `executionPayloadCapella`, `executionPayloadHeader` and `executionPayloadHeaderCapella`. The reason for it is to avoid duplication in functions such as this:
```
func (s *Service) notifyNewPayload(ctx context.Context, postStateVersion int,
	postStateHeader interfaces.ExecutionData, blk interfaces.SignedBeaconBlock) (bool, error)
```
Here `postStateHeader` is of type `ExecutionData`, which means a single function can handle both possible underlying header types.

**Which issues(s) does this PR fix?**

Part of #11077

### CODE REVIEW

File: beacon-chain/db/kv/state.go [hunk 1]
```diff
@@ -303,6 +303,30 @@ func (s *Store) saveStatesEfficientInternal(ctx context.Context, tx *bolt.Tx, bl
 			if err := valIdxBkt.Put(rt[:], validatorKeys[i]); err != nil {
 				return err
 			}
+		case *ethpb.BeaconStateCapella:
+			var pbState *ethpb.BeaconStateCapella
+			var err error
+			pbState, err = statenative.ProtobufBeaconStateCapella(rawType)
```
Comment (rkapka) [inline beacon-chain/db/kv/state.go:309]: Other states have `if features.Get().EnableNativeState` statements here, but I skipped it for Capella.

File: beacon-chain/state/state-native/hasher.go [hunk 2]
```diff
@@ -236,5 +238,30 @@ func ComputeFieldRootsWithHasher(ctx context.Context, state *BeaconState) ([][]b
...
+	if state.version == version.Capella {
+		// Execution payload root.
+		executionPayloadRoot, err := state.latestExecutionPayloadHeaderCapella.HashTreeRoot()
+		if err != nil {
+			return nil, err
+		}
+		fieldRoots[nativetypes.LatestExecutionPayloadHeaderCapella.RealPosition()] = executionPayloadRoot[:]
+
+		// Withdrawal queue root.
+		withdrawalQueueRoot := []byte("stub")
```
Comment (rkapka) [inline beacon-chain/state/state-native/hasher.go:250]: I left a gift for you @potuz

File: beacon-chain/state/state-native/getters_payload_header.go [hunk 3]
```diff
@@ -8,7 +8,7 @@ import (
 
 // LatestExecutionPayloadHeader of the beacon state.
 func (b *BeaconState) LatestExecutionPayloadHeader() (*enginev1.ExecutionPayloadHeader, error) {
-	if b.version == version.Phase0 || b.version == version.Altair {
+	if b.version != version.Bellatrix {
```
Comment (potuz) [inline beacon-chain/state/state-native/getters_payload_header.go:11]: Just a placeholder comment to look into if we create the interface now

File: beacon-chain/state/state-native/getters_payload_header.go [hunk 4]
```diff
@@ -27,3 +27,25 @@ func (b *BeaconState) LatestExecutionPayloadHeader() (*enginev1.ExecutionPayload
...
+	}
+
+	if b.latestExecutionPayloadHeaderCapella == nil {
+		return nil, nil
+	}
+
+	b.lock.RLock()
+	defer b.lock.RUnlock()
+
+	return b.latestExecutionPayloadHeaderCapellaVal(), nil
```
Comment (potuz) [inline beacon-chain/state/state-native/getters_payload_header.go:44]: Why the extra helper function that is a one liner?

File: beacon-chain/state/state-native/getters_withdrawal.go [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+	return ethpb.CopyWithdrawalSlice(b.withdrawalQueue)
+}
+
+// NextWithdrawalIndex returns the index that will be assigned to the next withdrawal.
+func (b *BeaconState) NextWithdrawalIndex() (uint64, error) {
+	if b.version < version.Capella {
+		return 0, errNotSupported("NextWithdrawalIndex", b.version)
+	}
+
+	b.lock.RLock()
```
Comment (potuz) [inline beacon-chain/state/state-native/getters_withdrawal.go:32]: do we want a lock to read the version?

File: beacon-chain/state/state-native/hasher.go [hunk 6]
```diff
@@ -236,5 +238,30 @@ func ComputeFieldRootsWithHasher(ctx context.Context, state *BeaconState) ([][]b
...
+		fieldRoots[nativetypes.WithdrawalQueue.RealPosition()] = withdrawalQueueRoot
+
+		// Next withdrawal index root.
+		nextWithdrawalIndexBuf := make([]byte, 8)
+		binary.LittleEndian.PutUint64(nextWithdrawalIndexBuf, state.nextWithdrawalIndex)
+		nextWithdrawalIndexRoot := bytesutil.ToBytes32(nextWithdrawalIndexBuf)
```
Comment (potuz) [inline beacon-chain/state/state-native/hasher.go:256]: Why the double copy here: can't we just put the 8 bytes directly in the 32 bytes buffer?

File: beacon-chain/db/kv/state.go [hunk 7]
```diff
@@ -311,6 +244,113 @@ func (s *Store) saveStatesEfficientInternal(ctx context.Context, tx *bolt.Tx, bl
 	return s.storeValidatorEntriesSeparately(ctx, tx, validatorsEntries)
 }
 
+func savePhase0Efficient(ctx context.Context, rawType interface{}, bucket *bolt.Bucket, blockRoot [32]byte, valIdxBkt *bolt.Bucket, valKey []byte) error {
```
Comment (rkapka) [inline beacon-chain/db/kv/state.go:247]: I extracted several functions because the complexity of the main func was too big for our CI check

File: beacon-chain/state/interfaces.go [hunk 8]
```diff
@@ -62,7 +61,7 @@ type ReadOnlyBeaconState interface {
 	MarshalSSZ() ([]byte, error)
 	IsNil() bool
 	Version() int
-	LatestExecutionPayloadHeader() (*enginev1.ExecutionPayloadHeader, error)
+	LatestExecutionPayloadHeader() (interfaces.ExecutionData, error)
```
Comment (rkapka) [inline beacon-chain/state/interfaces.go:64]: I decided to reuse the interface that Raul created for blocks.

File: beacon-chain/state/state-native/beacon_state_minimal.go [hunk 9]
```diff
@@ -18,34 +18,38 @@ import (
...
+	previousJustifiedCheckpoint         *ethpb.Checkpoint                       `ssz-gen:"true"`
+	currentJustifiedCheckpoint          *ethpb.Checkpoint                       `ssz-gen:"true"`
+	finalizedCheckpoint                 *ethpb.Checkpoint                       `ssz-gen:"true"`
+	inactivityScores                    []uint64                                `ssz-gen:"true" ssz-max:"1099511627776"`
+	currentSyncCommittee                *ethpb.SyncCommittee                    `ssz-gen:"true"`
+	nextSyncCommittee                   *ethpb.SyncCommittee                    `ssz-gen:"true"`
+	latestExecutionPayloadHeader        *enginev1.ExecutionPayloadHeader        `ssz-gen:"true"`
+	latestExecutionPayloadHeaderCapella *enginev1.ExecutionPayloadHeaderCapella `ssz-gen:"true"`
+	withdrawalQueue                     []*enginev1.Withdrawal                  `ssz-gen:"true" ssz-max:"1099511627776"`
+	nextWithdrawalIndex                 uint64                                  `ssz-gen:"true"`
```
[the remark this answers is not in the document]
Comment (rkapka) [reply beacon-chain/state/state-native/beacon_state_minimal.go:51]: We can definitely create a `WithdrawalIndex` type to avoid mistakes.

File: consensus-types/wrapper/execution.go [hunk 10]
```diff
@@ -293,6 +296,287 @@ func PayloadToHeader(payload interfaces.ExecutionData) (*enginev1.ExecutionPaylo
...
+		PrevRandao:       bytesutil.SafeCopyBytes(payload.PrevRandao()),
+		BlockNumber:      payload.BlockNumber(),
+		GasLimit:         payload.GasLimit(),
+		GasUsed:          payload.GasUsed(),
+		Timestamp:        payload.Timestamp(),
+		ExtraData:        bytesutil.SafeCopyBytes(payload.ExtraData()),
+		BaseFeePerGas:    bytesutil.SafeCopyBytes(payload.BaseFeePerGas()),
+		BlockHash:        bytesutil.SafeCopyBytes(payload.BlockHash()),
+		TransactionsRoot: txRoot[:],
+		WithdrawalsRoot:  []byte("stub"),
```
Comment (rkapka) [inline consensus-types/wrapper/execution.go:576]: Another exercise for the reader @potuz ;-)

File: beacon-chain/state/state-native/getters_withdrawal.go (hunk 5 above)
Comment (rkapka) [reply beacon-chain/state/state-native/getters_withdrawal.go:32]: Version? This is `NextWithdrawalIndex`. And we don't have a lock when reading the version.

File: consensus-types/blocks/execution.go [hunk 11]
```diff
@@ -293,6 +295,287 @@ func PayloadToHeader(payload interfaces.ExecutionData) (*enginev1.ExecutionPaylo
...
+		PrevRandao:       bytesutil.SafeCopyBytes(payload.PrevRandao()),
+		BlockNumber:      payload.BlockNumber(),
+		GasLimit:         payload.GasLimit(),
+		GasUsed:          payload.GasUsed(),
+		Timestamp:        payload.Timestamp(),
+		ExtraData:        bytesutil.SafeCopyBytes(payload.ExtraData()),
+		BaseFeePerGas:    bytesutil.SafeCopyBytes(payload.BaseFeePerGas()),
+		BlockHash:        bytesutil.SafeCopyBytes(payload.BlockHash()),
+		TransactionsRoot: txRoot[:],
+		WithdrawalsRoot:  []byte("stub"),
```
Comment (rkapka) [inline consensus-types/blocks/execution.go:575]: Another one for @potuz

File: beacon-chain/db/kv/state.go [hunk 12]
```diff
@@ -294,6 +294,28 @@ func (s *Store) saveStatesEfficientInternal(ctx context.Context, tx *bolt.Tx, bl
 			if err := valIdxBkt.Put(rt[:], validatorKeys[i]); err != nil {
 				return err
 			}
+		case *ethpb.BeaconStateCapella:
```
Comment (potuz) [inline beacon-chain/db/kv/state.go:297]: This case switch is a good candidate for simplification using generics since all cases are the same function based on pbState.

File: beacon-chain/state/state-native/beacon_state_minimal.go [hunk 13]
```diff
@@ -18,34 +18,38 @@ import (
...
-	inactivityScores             []uint64                         `ssz-gen:"true" ssz-max:"1099511627776"`
-	currentSyncCommittee         *ethpb.SyncCommittee             `ssz-gen:"true"`
-	nextSyncCommittee            *ethpb.SyncCommittee             `ssz-gen:"true"`
-	latestExecutionPayloadHeader *enginev1.ExecutionPayloadHeader `ssz-gen:"true"`
+	version                             int
+	genesisTime                         uint64                                  `ssz-gen:"true"`
+	genesisValidatorsRoot               customtypes.Byte32                      `ssz-gen:"true" ssz-size:"32"`
+	slot                                eth2types.Slot                          `ssz-gen:"true"`
+	fork                                *ethpb.Fork                             `ssz-gen:"true"`
+	latestBlockHeader                   *ethpb.BeaconBlockHeader                `ssz-gen:"true"`
```
Comment (potuz) [inline beacon-chain/state/state-native/beacon_state_minimal.go:26]: Will this change in Capella?

File: beacon-chain/state/state-native/getters_payload_header.go [hunk 14]
```diff
@@ -19,11 +21,20 @@ func (b *BeaconState) LatestExecutionPayloadHeader() (*enginev1.ExecutionPayload
 	b.lock.RLock()
 	defer b.lock.RUnlock()
 
-	return b.latestExecutionPayloadHeaderVal(), nil
+	if b.version == version.Bellatrix {
```
Comment (potuz) [inline beacon-chain/state/state-native/getters_payload_header.go:24]: I think in general and considering future forks, the best approach for all these switches is the opposite order
```
if version >= Capella { do something  }
else if version >= Bellatrix { do something }
else { not supported }
```
The first one is hit more often.

File: beacon-chain/state/state-native/state_trie.go [hunk 15]
```diff
@@ -731,6 +820,14 @@ func (b *BeaconState) rootSelector(ctx context.Context, field nativetypes.FieldI
 		return stateutil.SyncCommitteeRoot(b.nextSyncCommittee)
 	case nativetypes.LatestExecutionPayloadHeader:
 		return b.latestExecutionPayloadHeader.HashTreeRoot()
+	case nativetypes.LatestExecutionPayloadHeaderCapella:
+		return b.latestExecutionPayloadHeaderCapella.HashTreeRoot()
+	case nativetypes.WithdrawalQueue:
+		return [32]byte{}, errors.New("not implemented")
```
Comment (rkapka) [inline beacon-chain/state/state-native/state_trie.go:826]: and one more @potuz

File: beacon-chain/state/state-native/hasher.go (hunk 2 above)
Comment (potuz) [reply beacon-chain/state/state-native/hasher.go:250]: These are hashed as normal SSZ lists of `Withdrawal` objects, doesn't  fastssz already cover this?

File: beacon-chain/state/state-native/state_trie.go (hunk 15 above)
Comment (potuz) [reply beacon-chain/state/state-native/state_trie.go:826]: Same as above, why doesn't fastssz already cover this give it's a list?

File: consensus-types/blocks/execution.go (hunk 11 above)
Comment (potuz) [reply consensus-types/blocks/execution.go:575]: Same here this is the usual `list.HTR` from the withdrawal list in the execution payload.

File: beacon-chain/state/state-native/hasher.go (hunk 2 above)
Comment (potuz) [reply beacon-chain/state/state-native/hasher.go:250]: discussed offline, we can reuse Nishant's vectorized merkleizer

File: beacon-chain/state/state-native/beacon_state_minimal.go (hunk 13 above)
Comment (potuz) [reply beacon-chain/state/state-native/beacon_state_minimal.go:26]: discussed offline there's no change to the header.

File: beacon-chain/state/state-native/hasher.go [hunk 16]
```diff
@@ -236,5 +238,28 @@ func ComputeFieldRootsWithHasher(ctx context.Context, state *BeaconState) ([][]b
...
+		fieldRoots[nativetypes.LatestExecutionPayloadHeaderCapella.RealPosition()] = executionPayloadRoot[:]
+
+		// Withdrawal queue root.
+		withdrawalQueueRoot := []byte("stub")
+		fieldRoots[nativetypes.WithdrawalQueue.RealPosition()] = withdrawalQueueRoot
+
+		// Next withdrawal index root.
+		nextWithdrawalIndexRoot := make([]byte, 32)
+		binary.LittleEndian.PutUint64(nextWithdrawalIndexRoot, state.nextWithdrawalIndex)
+		fieldRoots[nativetypes.NextWithdrawalIndex.RealPosition()] = nextWithdrawalIndexRoot[:]
```
Comment (potuz) [inline beacon-chain/state/state-native/hasher.go:256]:
```suggestion
		fieldRoots[nativetypes.NextWithdrawalIndex.RealPosition()] = nextWithdrawalIndexRoot
```

File: beacon-chain/state/state-native/hasher.go [hunk 17]
```diff
@@ -236,5 +238,28 @@ func ComputeFieldRootsWithHasher(ctx context.Context, state *BeaconState) ([][]b
...
+
+		// Next partial withdrawal validator index root.
+		nextPartialWithdrawalValidatorIndexRoot := make([]byte, 32)
+		binary.LittleEndian.PutUint64(nextPartialWithdrawalValidatorIndexRoot, uint64(state.nextPartialWithdrawalValidatorIndex))
+		fieldRoots[nativetypes.NextPartialWithdrawalValidatorIndex.RealPosition()] = nextPartialWithdrawalValidatorIndexRoot[:]
```
Comment (potuz) [inline beacon-chain/state/state-native/hasher.go:261]:
```suggestion
		fieldRoots[nativetypes.NextPartialWithdrawalValidatorIndex.RealPosition()] = nextPartialWithdrawalValidatorIndexRoot
```

File: beacon-chain/state/state-native/beacon_state_minimal.go (hunk 13 above)
Comment (rkapka) [reply beacon-chain/state/state-native/beacon_state_minimal.go:26]: No :-)

File: beacon-chain/db/kv/state.go (hunk 12 above)
Comment (rkapka) [reply beacon-chain/db/kv/state.go:297]: Yes, this was my impression too. But I didn't want to refactor existing code in this PR.

Comment (rauljordan) [reply beacon-chain/db/kv/state.go:297]: There isn't even a need for generics if we can just pass in fastssz.Marshaler as an interface (beacon state satisfies this)

File: beacon-chain/state/state-native/hasher.go (hunk 2 above)
Comment (nisdas) [reply beacon-chain/state/state-native/hasher.go:250]: Do we intend to fix this ? Or is this meant to be in a follow up PR. i rather not have our hashing incomplete like this if we are adding in the whole state.

File: beacon-chain/state/state-native/setters_payload_header.go [hunk 18]
```diff
@@ -14,14 +14,20 @@ func (b *BeaconState) SetLatestExecutionPayloadHeader(val interfaces.ExecutionDa
 	b.lock.Lock()
 	defer b.lock.Unlock()
 
-	if b.version == version.Phase0 || b.version == version.Altair {
+	if b.version < version.Bellatrix {
 		return errNotSupported("SetLatestExecutionPayloadHeader", b.version)
 	}
 	header, ok := val.Proto().(*enginev1.ExecutionPayloadHeader)
```
Comment (nisdas) [inline beacon-chain/state/state-native/setters_payload_header.go:20]: can we use a switch statement here rather than multiple type assertions. Would make the code cleaner to read

File: beacon-chain/state/state-native/state_trie.go [hunk 19]
```diff
@@ -451,9 +540,7 @@ func (b *BeaconState) Copy() state.BeaconState {
 	switch b.version {
 	case version.Phase0:
 		dst.sharedFieldReferences = make(map[nativetypes.FieldIndex]*stateutil.Reference, 10)
-	case version.Altair:
-		dst.sharedFieldReferences = make(map[nativetypes.FieldIndex]*stateutil.Reference, 11)
-	case version.Bellatrix:
+	case version.Altair, version.Bellatrix, version.Capella:
```
Comment (nisdas) [inline beacon-chain/state/state-native/state_trie.go:543]: this isnt correct, altair, bellatrix and capella have different sizes.

File: beacon-chain/state/state-native/state_trie.go (hunk 15 above)
Comment (nisdas) [reply beacon-chain/state/state-native/state_trie.go:826]: its not hard to do custom hashing for this too.

File: beacon-chain/state/state-native/getters_withdrawal.go (hunk 5 above)
Comment (rauljordan) [reply beacon-chain/state/state-native/getters_withdrawal.go:32]: I think @potuz refers to `b.version < version.Capella`. This version field is something that could be updated while another thread could be trying to read it

File: beacon-chain/blockchain/execution_engine.go [hunk 20]
```diff
@@ -183,7 +183,7 @@ func (s *Service) getPayloadHash(ctx context.Context, root []byte) ([32]byte, er
 // notifyForkchoiceUpdate signals execution engine on a new payload.
 // It returns true if the EL has returned VALID for the block
 func (s *Service) notifyNewPayload(ctx context.Context, postStateVersion int,
-	postStateHeader *enginev1.ExecutionPayloadHeader, blk interfaces.SignedBeaconBlock) (bool, error) {
+	postStateHeader interfaces.ExecutionData, blk interfaces.SignedBeaconBlock) (bool, error) {
```
Comment (terencechain) [inline beacon-chain/blockchain/execution_engine.go:186]: What's the reason for changing `*enginev1.ExecutionPayloadHeader` to `interfaces.ExecutionData`. First glance, it seems unrelated to the PR title. Can you please explain it or even better cover that in the description?

File: beacon-chain/state/state-native/setters_withdrawal.go [hunk 21]
```diff
@@ -0,0 +1,86 @@
...
+}
+
+// SetNextWithdrawalIndex sets the index that will be assigned to the next withdrawal.
+func (b *BeaconState) SetNextWithdrawalIndex(i uint64) error {
+	if b.version < version.Capella {
+		return errNotSupported("SetNextWithdrawalIndex", b.version)
+	}
+
+	b.lock.RLock()
+	defer b.lock.RUnlock()
```
Comment (terencechain) [inline beacon-chain/state/state-native/setters_withdrawal.go:68]: Should we move these read locks before the version check? Currently, it's not consistent with `SetWithdrawalQueue` and the setter method

**Review by terencechain — CHANGES_REQUESTED** [review]

File: beacon-chain/blockchain/execution_engine.go (hunk 20 above)
Comment (rkapka) [reply beacon-chain/blockchain/execution_engine.go:186]: I updated the description

File: beacon-chain/state/interfaces.go [hunk 22]
```diff
@@ -85,6 +85,10 @@ type WriteOnlyBeaconState interface {
 	UpdateSlashingsAtIndex(idx, val uint64) error
 	AppendHistoricalRoots(root [32]byte) error
 	SetLatestExecutionPayloadHeader(payload interfaces.ExecutionData) error
+	SetWithdrawalQueue(val []*enginev1.Withdrawal) error
```
[the remark this answers is not in the document]
Comment (rkapka) [reply beacon-chain/state/interfaces.go:88]: See discussion starting in https://discord.com/channels/476244492043812875/998988600530194583/1001913104512335882

**Review by terencechain — DISMISSED** [review]

**Comment by terencechain** [comment]
Let's get one more approve before merge plz

**Review by rauljordan — APPROVED** [review]

File: beacon-chain/state/state-native/beacon_state_mainnet.go [hunk 23]
```diff
@@ -18,34 +18,38 @@ import (
...
+	currentEpochParticipation           []byte                                  `ssz-gen:"true" ssz-max:"1099511627776"`
+	justificationBits                   bitfield.Bitvector4                     `ssz-gen:"true" ssz-size:"1"`
+	previousJustifiedCheckpoint         *ethpb.Checkpoint                       `ssz-gen:"true"`
+	currentJustifiedCheckpoint          *ethpb.Checkpoint                       `ssz-gen:"true"`
+	finalizedCheckpoint                 *ethpb.Checkpoint                       `ssz-gen:"true"`
+	inactivityScores                    []uint64                                `ssz-gen:"true" ssz-max:"1099511627776"`
+	currentSyncCommittee                *ethpb.SyncCommittee                    `ssz-gen:"true"`
+	nextSyncCommittee                   *ethpb.SyncCommittee                    `ssz-gen:"true"`
+	latestExecutionPayloadHeader        *enginev1.ExecutionPayloadHeader        `ssz-gen:"true"`
+	latestExecutionPayloadHeaderCapella *enginev1.ExecutionPayloadHeaderCapella `ssz-gen:"true"`
```
[the remark this answers is not in the document]
Comment (Inphi) [reply beacon-chain/state/state-native/beacon_state_mainnet.go:49]: The interface's method set could be that super-set though. We ran into the same issue updating the payload for eip-4844 and have gone with the [interface approach](https://github.com/Inphi/prysm/blob/eb1b85d841bce912d67062bbb6fefc9aa4ae31af/beacon-chain/state/state-native/beacon_state_mainnet.go#L20). This avoids bloating the beacon state with more and more fields as the the Execution Payload is updated in the future.
